File: python/pe014.py
cache = {}          #dynamic programming
def collatz(n):     #not using recursing because it is expensive
    original_n = int(eval(str(n)))
    c = []
    c.append(n)
    while n != 1:
        if n % 2 == 0:
            n = int(n/2)
            try:
                return len(c + cache[n])
            except:
                c.append(n)
        else:
            n = 3 * n + 1
            try:
                return len(c + cache[n])
            except:
                c.append(n)
    cache[original_n] = c
    return len(c)  
a = list(map(collatz, list(range(500000,1000000))))    #starting at 500,000 was a guess
maxi = 0
maximum = 0
for i in range(500000-1):
    if a[i] > maximum:
        maximum = a[i]
        maxi = i+500000
ans = maxi
print(ans)

File: cpp/pe014.cc
#include <iostream>
#include <unordered_map>

typedef long long int LL;

int collatz(const LL n, std::unordered_map<LL, int>& cache) {
  if (n == 1) return 1;
  if (cache.find(n) == cache.end()) {
    if (n % 2 == 0) {
      cache[n] = 1 + collatz(n/2, cache);
    } else {
      cache[n] = 1 + collatz(3*n+1, cache);
    }
  }
  return cache[n];
}

int main() {
  std::unordered_map<LL, int> cache;
  LL m = 0;
  int mn = 0;

  for (LL n = 1; n < 1000000; ++n) {
    int ans = collatz(n, cache);
    if (ans > mn) {
      m = n;
      mn = ans;
    }
  }

  std::cout << m << " " << mn << std::endl;
}
